### PR TITLE
Docker development environment

### PR DESCRIPTION
**Pull request by peter-edge**
I will try to do a readme at some point, but for now:

```
git fetch origin && git checkout docker-dev
make help
make ci
export SUPPRESS_DOCKER=1
make help
make ci
unset SUPPRESS_DOCKER
```

### CODE REVIEW

**Comment by willhug** [comment]
This is a bit daunting to review without context about what is changing and why.  Can you add that readme/message describing:
- What changed
- Why we changed it
- How I can debug/modify this in the future (maybe with some links)

File: build/crossdock.mk [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+DOCKER_COMPOSE_ARCH := $(UNAME_ARCH)
+
+DOCKER_COMPOSE = $(BIN)/docker-compose
+
+$(DOCKER_COMPOSE):
+	@mkdir -p $(BIN)
+	curl -L https://github.com/docker/compose/releases/download/$(DOCKER_COMPOSE_VERSION)/docker-compose-$(DOCKER_COMPOSE_OS)-$(DOCKER_COMPOSE_ARCH) > $(DOCKER_COMPOSE)
+	@chmod +x $(DOCKER_COMPOSE)
+
+.PHONY: crossdock
```
Comment (breerly) [inline build/crossdock.mk]: Organizing all the crossdock targets in this file is nice.

**Comment by peter-edge** [comment]
We can video chat and I can explain, I am way too tired right now to write
up a whole readme, what do you think?

**Comment by peter-edge** [comment]
Why: https://github.com/yarpc/yarpc-go/issues/826

Except this PR actually keeps local in tact.

File: build/deps.mk [hunk 2]
```diff
@@ -0,0 +1,107 @@
+# List of vendored go executables needed for make generate
+GEN_GO_BIN_DEPS = \
+	github.com/golang/mock/mockgen \
+	github.com/gogo/protobuf/protoc-gen-gogoslick \
+	github.com/uber/tchannel-go/thrift/thrift-gen \
+	golang.org/x/tools/cmd/stringer \
+	go.uber.org/thriftrw \
```
Comment (breerly) [inline build/deps.mk]: I'm always nervous when it comes to installing bins from master...

Makes me wonder if we can get these things using glide and then install them directly out of `vendor/` so we can control the versions explicitly.

**Comment by peter-edge** [comment]
I would encourage playing with this locally as a good way to check it out.

**Comment by breerly** [comment]
Also, I wonder if the transition might be easier if we default SUPPRESS_DOCKER to 1.

**Comment by peter-edge** [comment]
Back at my computer in a sec - note I've been combing Travis builds all
day, and I can confirm  it works. Also I never ever want to hear "Travis
CI" ever again.

I would encourage ppl adding that to their bashrc instead - having docker
as the default is what I want to move to, esp for OSS engineers who come
into this externally from uber.

File: .dockerignore [hunk 3]
```diff
@@ -6,6 +6,7 @@
 *.t.err
 .tmp
 .cache
+.glide
```
Comment (peter-edge) [inline .dockerignore]: Check the generate script, I already am :)

File: .travis.yml [hunk 4]
```diff
@@ -1,35 +1,35 @@
...
-        - CI_TYPES='examples lint test'
-
+      sudo: required
+      services:
+        - docker
+    - env: CI_TYPES='goveralls' SUPPRESS_CROSSDOCK=1 DOCKER_GO_VERSION=1.7
+      sudo: required
+      services:
+        - docker
+    - env: CI_TYPES='lint test' SUPPRESS_CROSSDOCK=1 DOCKER_GO_VERSION=1.8
```
Comment (peter-edge) [inline .travis.yml]: No, see below

File: build/deps.mk (hunk 2 above)
Comment (peter-edge) [inline build/deps.mk]: That IS what is happening :)

**Comment by peter-edge** [comment]
Making one more fix again - travis has gone back and forth TODAY from docker-compose being called `yarpcgo_go` to and from `yarpc_go`.

File: build/deps.mk (hunk 2 above)
Comment (breerly) [inline build/deps.mk]: *doh*

File: .dockerignore (hunk 3 above)
Comment (breerly) [inline .dockerignore]: *doh*

**Comment by breerly** [comment]
Ok, I just test drove this locally, and it is freaking sweet.

@peter-edge I'm fine with landing this with the 1 stamp.

The real value will come from having the team adopt - of course we can make adjustments along the way.

**Comment by peter-edge** [comment]
Ya, thanks for the support dudee.

**Pull request merged**